scripts/tools/zapfile_formatter.py

```diff
@@ -0,0 +1,199 @@
+#    limitations under the License.
+#
+
+import argparse
+import json
+from typing import List
+
+_DEFAULT_CLUSTER_REVISION_ATTRIBUTE = {
+    "name": "ClusterRevision",
+    "code": 0xFFFD,
+    "mfgCode": None,
+    "side": "server",
+    "type": "int16u",
+    "included": 1,
+    "storageOption": "RAM",
+    "singleton": 0,
+    "bounded": 0,
+    "defaultValue": "1",
+    "reportable": 1,
+    "minInterval": 1,
+    "maxInterval": 65534,
+    "reportableChange": 0
+}
+
+_DEFAULT_FEATURE_MAP_ATTRIBUTE = {
+    "name": "FeatureMap",
+    "code": 0xFFFC,
+    "mfgCode": None,
+    "side": "server",
+    "type": "bitmap32",
+    "included": 1,
+    "storageOption": "RAM",
+    "singleton": 0,
+    "bounded": 0,
+    "defaultValue": "0",
+    "reportable": 1,
+    "minInterval": 1,
+    "maxInterval": 65534,
+    "reportableChange": 0
+}
+
+
+class Mutator:
+    def __init__(self):
+        pass
+
+    def handle(self, candidate: object):
+        pass
+
+
+class ValidateManditoryServerClusterAttributes(Mutator):
+    def __init__(self, attribute_entry, add_if_missing, forces_include, replace_if_storage_nvm):
+        self._attribute_entry = attribute_entry
+        self._add_if_missing = add_if_missing
+        self._forces_include = forces_include
+        self._replace_if_storage_nvm = replace_if_storage_nvm
+        super().__init__()
+
+    def _addMissingManditoryAttribute(self, candidate: object):
+        attributes = candidate.get("attributes", [])
+        insert_index = 0
+        for attribute in attributes:
+            attribute_code = attribute.get("code")
+            if attribute_code is not None and attribute_code > self._attribute_entry[
+                    "code"]:
+                break
+
+            insert_index += 1
+
+        # Insert the new attribute in the right place, WITH NO RENUMBERING
+        new_attrib_list = attributes[0:insert_index]
+        new_attrib_list.append(self._attribute_entry)
+        new_attrib_list.extend(attributes[insert_index:])
+
+        # Replace the attribute list with the augmented item
+        candidate["attributes"] = new_attrib_list
+
+    def handle(self, candidate: object):
+        if not isinstance(candidate, dict):
+            return
+
+        # We only care about adding manditory attributes.
+        if "attributes" not in candidate:
+            return
+
+        # If the cluster is not a server or is not enabled we do not enforce adding manidory attribute.
+        if (("enabled" not in candidate) or ("side" not in candidate)):
+            return
+        if (not candidate.get("enabled")) or ("server"
+                                              not in candidate.get("side")):
+            return
+
+        for attribute in candidate.get("attributes", []):
+            if attribute["code"] != self._attribute_entry["code"]:
+                continue
+            if attribute["name"] != self._attribute_entry["name"]:
```

Comment:
The name in the .zap file is basically ignored by everything.  It's there just to make it a bit more readable, I think.  We should ignore it too, imo.